#365 systeminfo binding

svilenvul opened the pull request:
Pull request for #365 

As currently there is no support for dynamic channels at DSL defined things [(see](https://github.com/eclipse/smarthome/issues/588)), I have not implemented this feature. 
Once the issue is resolved, I can update the binding. 

Signed-off-by: Svilen Valkanov <svilen.valkanov@musala.com>     (github: svilenvul)

watou commented:
Thanks for doing this!

Would you mind commenting on my PR on the 1.x systeminfo binding here https://github.com/openhab/openhab/pull/4183?

It might be relevant to your 2.x binding.  Thanks.

hakan42 commented:
Strangely, I cannot add comments on the review tab. @kaikreuzer , I should be member of openhab2-addons as well, shouldn't I?

hakan42 commented:
In addons/binding/org.openhab.binding.systeminfo/META-INF/MANIFEST.MF , do you need the specific version of commons-lang? If not, could you remove the version information?

hakan42 commented:
In addons/binding/pom.xml , could you please check your tabs-and-spaces settings?

kaikreuzer commented:
> Strangely, I cannot add comments on the review tab.

Hm, I would expected so as well - you are in the list of users to which issues can be assigned, etc. (but without having repo write access).

hakan42 commented:
Makes reviewing a tiny bit harder, is all I'm saying :smiling_imp:

svilenvul commented:
@hakan42, thanks I will take care

kaikreuzer commented on addons/binding/org.openhab.binding.systeminfo.test/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,26 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Fragment-Host: org.openhab.binding.systeminfo
+Bundle-Name: Systeminfo Test Binding
```
better: "Systeminfo Binding Tests"

kaikreuzer commented on addons/binding/org.openhab.binding.systeminfo/ESH-INF/thing/computer.xml:
```diff
@@ -0,0 +1,56 @@
+		<description>Sample thing for systeminfo Binding</description>
+
+		<channels>
+			<channel id="os_manufacturer" typeId="os_manufacturer"></channel>
```
how about using channel groups? This binding seems to be predestined for that :-)

kaikreuzer commented on addons/binding/org.openhab.binding.systeminfo/README.md:
```diff
@@ -0,0 +1,162 @@
+
+## Discovery
+
+Discovery is not necessary.
```
Actually, you could think about always discovering the local computer, right?

kaikreuzer commented on addons/binding/org.openhab.binding.systeminfo/src/main/java/org/openhab/binding/systeminfo/handler/SysteminfoHandler.java:
```diff
@@ -0,0 +1,379 @@
+
+    @Override
+    public void initialize() {
+        System.out.println("Start init.");
```
please remove all System.outs in here :-)

kaikreuzer commented on addons/binding/org.openhab.binding.systeminfo/src/main/java/org/openhab/binding/systeminfo/handler/SysteminfoHandler.java:
```diff
@@ -0,0 +1,379 @@
+        try {
+            this.systeminfo = new OshiSysteminfo();
+        } catch (SocketException e) {
+            logger.error("Network information is not availble ! I/O error occured {e}", e);
```
typo: available

kaikreuzer commented on addons/binding/org.openhab.binding.systeminfo/src/main/java/org/openhab/binding/systeminfo/model/DeviceNotFoundException.java:
```diff
@@ -0,0 +1,65 @@
+     *            {@link #getCause()} method). (A null value is permitted,
+     *            and indicates that the cause is nonexistent or unknown.)
+     *
+     * @since 1.6
```
please remove the since annotation

kaikreuzer commented on addons/binding/pom.xml:
```diff
@@ -46,7 +46,9 @@
     <module>org.openhab.binding.yamahareceiver</module>
     <module>org.openhab.binding.zwave</module>
+    <module>org.openhab.binding.systeminfo</module>
+	<module>org.openhab.binding.systeminfo.test</module>
```
please replace the tab by spaces

kaikreuzer commented:
Hi @svilenvul!

Very nice code and a cool feature - it is great that this does not rely anymore on native code :-)
My main wish would be the introduction of channel groups (including marking quite some channels as "advanced"), to make it a less long list in the UI. You might also want to think about a simple DiscoveryService.

svilenvul commented:
Thank you @kaikreuzer for your comments.

I think I have corrected the above mentioned problems. Channel groups, "advanced" channels, discovery service and some new functionality are now implemented.

Please let me know, if I have missed to fix something :)

kaikreuzer commented:
Thanks @svilenvul!

I just tested it and came across a problem in the discovery service:
```
2016-05-08 23:23:23.814 [ERROR] [i.DiscoveryServiceRegistryImpl:339  ] - Cannot trigger scan for thing types '[systeminfo:computer]' on 'SysteminfoDiscoveryService'!java.lang.IllegalArgumentException: UID segment 'Hilo.fritz.box' contains invalid characters. Each segment of the UID must match the pattern [A-Za-z0-9_-]*.
	at org.eclipse.smarthome.core.thing.UID.validateSegment(UID.java:91)
	at org.eclipse.smarthome.core.thing.UID.<init>(UID.java:69)
	at org.eclipse.smarthome.core.thing.ThingUID.<init>(ThingUID.java:40)
	at org.openhab.binding.systeminfo.discovery.SysteminfoDiscoveryService.startScan(SysteminfoDiscoveryService.java:61)
	at org.eclipse.smarthome.config.discovery.AbstractDiscoveryService.startScan(AbstractDiscoveryService.java:199)
	at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScan(DiscoveryServiceRegistryImpl.java:336)
	at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScans(DiscoveryServiceRegistryImpl.java:321)
	at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScan(DiscoveryServiceRegistryImpl.java:172)
	at org.eclipse.smarthome.io.rest.core.discovery.DiscoveryResource.scan(DiscoveryResource.java:79)
```
You need to make sure that the UID only uses allowed characters.

kaikreuzer commented:
The channel groups make it very nice in the Paper UI now, thanks!
Regarding "advanced" channels, I would suggest to be even more rigorous and only have 5-8 channels non-advanced (stuff like %CPU, %Memory, %Battery, etc.). Btw, did you think about adding static information as Thing properties only? Information like number of cores etc. will never ever change and thus status events do not really make sense, do they?

kaikreuzer commented on addons/binding/org.openhab.binding.systeminfo/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,381 @@
+		<config-description uri="systeminfo:channels:lowpriority" />
+	</channel-type>
+
+	<channel-type id="phisycalCores">
```
typo! Should be physicalCores

svilenvul commented:
@kaikreuzer 
The latest updates of the Systeminfo binding are already here.

> Btw, did you think about adding static information as Thing properties only? Information like number of cores etc. will never ever change and thus status events do not really make sense, do they?

I was not sure, which information should I add as a properties and added only the number of cpu cores. Should I include properties for OS information as well ?

The Jenkins build has failed with 
> ERROR: Maven JVM terminated unexpectedly with exit code 137

Is there a way to rerun it ?

kaikreuzer commented on addons/binding/org.openhab.binding.systeminfo/ESH-INF/thing/channels.xml:
```diff
@@ -265,48 +263,32 @@
-
-	<channel-type id="logicalCores">
+
+	<channel-type id="loadAvarage" advanced="true">
```
just found another typo: loadAverage

kaikreuzer commented:
> The latest updates of the Systeminfo binding are already here.

Thanks!

> Should I include properties for OS information as well ?

I would think so - as it is also static and won't change (at least not during runtime).
To make sure that the information is not lost, but available to the user, I just created https://github.com/eclipse/smarthome/issues/1492.

> Is there a way to rerun it ?

I don't really now. But never mind, if your build works locally, it is enough.

kaikreuzer commented:
Many thanks @svilenvul!
